src/hotspot/share/oops/arrayOop.hpp

```diff
@@ -85,7 +91,8 @@ class arrayOopDesc : public oopDesc {
 
   // Returns the offset of the first element.
   static int base_offset_in_bytes(BasicType type) {
-    return header_size(type) * HeapWordSize;
+    size_t hs = header_size_in_bytes();
```

Comment:
Previously, the memory representation of an array is `[header] + [payload]`. The "header" part includes the necessary alignment gap if any. (IOW, the end of the header and the beginning of the payload is adjacent.) The public API has `header_size(...)` and `base_offset_in_bytes(...)`, which corresponds to the header and payload, respectively.

With the new change, the header is *not* word-alignment any more. Could the API (and its impl) be adjusted to maintain the original model `array = [header] + [payload]`. Sth like, `header_size_in_bytes(...)` includes the alignment gap and replaces `header_size` in the public API.

Reply:
I am not sure what you are asking here. We *do* have header_size_in_bytes(), but its meaning is just the header, without gap (if any). The gap is included in base_offset_in_bytes(), which is the offset at which the payload starts, which seems consistent to other similar methods. We could make header_size_in_bytes() so that it includes the gap, but then 1. it would be the same as base_offset_in_bytes() and 2. we still need 'just the header' in places, so we would have to have another method for that. Also, I am not sure where we would need header_size_in_bytes() with gap, because that's usually the places where base_offset_in_bytes() seems semantically better suited.

(Also, note that we still have header_size(BasicType) which returns an word-up-aligned header size which is used by a number of places that require this like minimum and maximum size computations; it's kinda ugly, but having to have to align-up the byte-size in all those places has been ugly as well, I have attempted that in an earlier version of this PR)

Reply:
On master, `header_size` and `base_offset_in_bytes` are two public API; they refer to the header and payload of an array and have the invariant, header-size-in-bytes == base-offset-in-bytes, because `base_offset_in_bytes(...)` is impled as `header_size(...) * HeapWordSize`.

With the new change, `header_size` and `base_offset_in_bytes` are still two public API, but the invariant is broken, `header_size(...) * HeapWordSize != base_offset_in_bytes(...)`.  I am asking if we can expose `header_size_in_bytes` as the public API so that invariant is kept.

> 1. it would be the same as base_offset_in_bytes()

True; the impl on master is just a word-to-byte translation. (They being the same is an advantage, reinforcing the invariant, IMO.)

> 2. we still need 'just the header' in places

Is that a public API? I think the concept of "header" contains everything from array-start to payload-start, from the perspective of users (of this class).

> it's kinda ugly, but having to have to align-up the byte-size in all those places has been ugly as well

I am more concerned about the fact that the public API `header_size` covers more than the header. Having the header-size aligned up to word-size (or anything) should be sth the caller decides.

Reply:
> On master, `header_size` and `base_offset_in_bytes` are two public API; they refer to the header and payload of an array and have the invariant, header-size-in-bytes == base-offset-in-bytes, because `base_offset_in_bytes(...)` is impled as `header_size(...) * HeapWordSize`.
> 
> With the new change, `header_size` and `base_offset_in_bytes` are still two public API, but the invariant is broken, `header_size(...) * HeapWordSize != base_offset_in_bytes(...)`. I am asking if we can expose `header_size_in_bytes` as the public API so that invariant is kept.

So you want:
  static int base_offset_in_bytes() {
    return header_size_in_bytes();
  }
??
Seems like an unnecessary duplication to me.

> > 1. it would be the same as base_offset_in_bytes()
> 
> True; the impl on master is just a word-to-byte translation. (They being the same is an advantage, reinforcing the invariant, IMO.)

Yes, but that distinction is gone. We shall use only the byte-size version, except for the case where we really want the up-aligned word-sized version (this needs to be used with caution, see below).

> > 2. we still need 'just the header' in places
> 
> Is that a public API? I think the concept of "header" contains everything from array-start to payload-start, from the perspective of users (of this class).

No. All users are code-generators (except for base_offset_in_bytes() and header_size() which transform the value in their own ways). And we are using the header-size *without* the gap in order to determine if we need an extra store to initialize the gap. E.g.:

    if (!is_aligned(arrayOopDesc::header_size_in_bytes(), BytesPerWord)) {
      assert(is_aligned(arrayOopDesc::header_size_in_bytes(), BytesPerInt), "must be 4-byte aligned");
      strw(zr, Address(obj, arrayOopDesc::header_size_in_bytes()));
    }


> > it's kinda ugly, but having to have to align-up the byte-size in all those places has been ugly as well
> 
> I am more concerned about the fact that the public API `header_size` covers more than the header. Having the header-size aligned up to word-size (or anything) should be sth the caller decides.

Ok I agree. I've done that, and ended up with a dozen (or so) cases which all did the exact same thing. That's why I added back header_size() (I originally removed it) as a helper method for the callers which decided to do that. I'm open to suggestions for improving the situation.

Reply:
> And we are using the header-size without the gap in order to determine if we need an extra store to initialize the gap. E.g.:

I suspect the store to the gap is related to the following comment/diff:

```
  // We align-up the header size to word-size, because we clear the
  // possible alignment gap in initialize_header().
  int hdr_size = align_up(base_offset_in_bytes, BytesPerWord);
  initialize_body(obj, arr_size, hdr_size, len_zero);
```

Wonder if sth like `initialize_body(obj, arr_size, base_offset_in_bytes, len_zero);` can make the header/body classification more explicit. Ofc, `zero_memory` needs to be extended somehow to support non-word-align start-addr.

Reply:
> > And we are using the header-size without the gap in order to determine if we need an extra store to initialize the gap. E.g.:
> 
> I suspect the store to the gap is related to the following comment/diff:
> 
> ```
>   // We align-up the header size to word-size, because we clear the
>   // possible alignment gap in initialize_header().
>   int hdr_size = align_up(base_offset_in_bytes, BytesPerWord);
>   initialize_body(obj, arr_size, hdr_size, len_zero);
> ```
> 
> Wonder if sth like `initialize_body(obj, arr_size, base_offset_in_bytes, len_zero);` can make the header/body classification more explicit. Ofc, `zero_memory` needs to be extended somehow to support non-word-align start-addr.

I did that in an earlier version of this PR and it looked quite messy (you can see it if you go back in the commit history. The trouble is that the loops that zero the body are optimized to work on 8-byte-word units, so we would have to check if the body is aligned or not, and if not, add an extra store. I thought it was easier to just add that store in the header zeroing routine. This is also consistent with how we store the klass-gap in instance objects there.

Reply:
> I did that in an earlier version of this PR and it looked quite messy ...

I just checked the first commit and the `MacroAssembler` part matches what I have in mind. I believe that diff being messy is due to "Calculations of min and max array and tlab sizes".

> so we would have to check if the body is aligned or not, and if not, add an extra store

The store, as I understand it, belongs to the body (payload) semantically. Mixing stores btw `initialize_header` and `initialize_body` is super confusing, IMO.

> This is also consistent with how we store the klass-gap in instance objects there.

klass-gap is part of the header, so it makes sense to be in `initialize_header`.

Reply:
> > I did that in an earlier version of this PR and it looked quite messy ...
> 
> I just checked the first commit and the `MacroAssembler` part matches what I have in mind. I believe that diff being messy is due to "Calculations of min and max array and tlab sizes".

Yes, that is another part of it. I can re-instate the parts where we initialize the leading 4 bytes in the body, if agree that this is the way to go.
 
> > so we would have to check if the body is aligned or not, and if not, add an extra store
> 
> The store, as I understand it, belongs to the body (payload) semantically. Mixing stores btw `initialize_header` and `initialize_body` is super confusing, IMO.


Yes, I understand that.
 
> > This is also consistent with how we store the klass-gap in instance objects there.
> 
> klass-gap is part of the header, so it makes sense to be in `initialize_header`.

No, klass-gap really is not part of the header. For non-array objects, this gap is where the first bunch of fields go, it really is payload. So, arguably, we should move the clearing of the gap into initialize_body(), too - but not as part of this PR.

Reply:
There is sth odd about `instanceOopDesc` that the header/payload reported by public API `header_size` and `base_offset_in_bytes` can be overlapping, shown picturally:

```
|<-header->|....     |
|        |<-payload->|
```

Essentially, `header-size * HeapWordSize != base-offset-in-bytes`, blurring the boundary btw header and payload.

Possibly, it's intentional to overapproximate "header" and underapproximate "payload", because some context dependent constraints prevent `header_size` from being klass specific, i.e. having knowledge to the corresponding fields (layout). Otherwise, one should avoid declaring the existence of "klass-gap" then filling it up using some fields from the object. A klass-gap is there if and only if the first field has special alignment requirement.

> No, klass-gap really is not part of the header. For non-array objects, this gap is where the first bunch of fields go, it really is payload.

I can see where that interpretation comes from since `base_offset_in_bytes == klass_gap_offset_in_bytes` when `UseCompressedClassPointers == true` and base-offset denotes the start of payload.

It's not part of this diff -- in `C1_MacroAssembler::initialize_header`, can `movl(Address(obj, oopDesc::klass_offset_in_bytes()), t1);` use `movptr` like its 32bit counterpart?

Since the klass-layout-info is unavailable in this context, including this "potential" gap as part of the header is kind of overapproximation.

Reply:
> There is sth odd about `instanceOopDesc` that the header/payload reported by public API `header_size` and `base_offset_in_bytes` can be overlapping, shown picturally:
> 
> ```
> |<-header->|....     |
> |        |<-payload->|
> ```
> 
> Essentially, `header-size * HeapWordSize != base-offset-in-bytes`, blurring the boundary btw header and payload.
> 
> Possibly, it's intentional to overapproximate "header" and underapproximate "payload", because some context dependent constraints prevent `header_size` from being klass specific, i.e. having knowledge to the corresponding fields (layout). Otherwise, one should avoid declaring the existence of "klass-gap" then filling it up using some fields from the object. A klass-gap is there if and only if the first field has special alignment requirement.

With this patch, header_size() is really only a helper-method that aligns-up header_size_in_bytes() to the next word boundary, and returns that size in word-units, for use by code that need such a thing (e.g. various min/max size calculations that operate with heap-words). The source of truth regarding header-size vs payload-offset is the two methods header_size_in_bytes() and base_offset_in_bytes(). Maybe it makes sense to move out header_size() to avoid the confusion?

> > No, klass-gap really is not part of the header. For non-array objects, this gap is where the first bunch of fields go, it really is payload.
> 
> I can see where that interpretation comes from since `base_offset_in_bytes == klass_gap_offset_in_bytes` when `UseCompressedClassPointers == true` and base-offset denotes the start of payload.
> 
> It's not part of this diff -- in `C1_MacroAssembler::initialize_header`, can `movl(Address(obj, oopDesc::klass_offset_in_bytes()), t1);` use `movptr` like its 32bit counterpart?

It could, but it doesn't seem to be a good fit because a narrowKlass is not a ptr. narrowKlass are explicitly only
a 32bit int value and should be treated as such.

> Since the klass-layout-info is unavailable in this context, including this "potential" gap as part of the header is kind of overapproximation.

Yeah. If the gap is not used by instance fields or array elements, they don't actually belong to neither the header nor to the payload. It's just an alignment gap that sits in between the two. Strictly speaking they don't even need to be cleared (and I believe C2 would not clear it in any case) because it's not accessed, ever. But we don't have that info for instances, so we clear it anyway. *If* the gap is used (and that is common), it's clearly payload.

Reply:
> Maybe it makes sense to move out header_size() to avoid the confusion?

I believe so. Having header-bytes aligned up to word sounds like sth the caller interests, not callee.

> It could, but it doesn't seem to be a good fit because a narrowKlass is not a ptr. narrowKlass are explicitly only
a 32bit int value and should be treated as such.

That would get rid of the trailing zero-gap later, right?

```cpp
  else if (UseCompressedClassPointers) {
    xorptr(t1, t1);
    store_klass_gap(obj, t1);
  }
```

> If the gap is not used by instance fields or array elements, they don't actually belong to neither the header nor to the payload.

It boils down to the def. of "header".

Option 1: `header == markword + klass + (length)         ==> array/obj == header + (gap) + payload/body`
Option 2: `header == markword + klass + (length) + (gap) ==> array/obj == header + payload/body`

It's kind of a preexisting issue though.

> and I believe C2 would not clear it in any case

I used a simple class to test.

```
class A {
  long x;
}
```

In the C2 assembly, I believe the following three instructions correspond to markword, klass and the gap.

```
  0x00007f7eccad9c60:   mov    QWORD PTR [rax],0x1
  0x00007f7eccad9c67:   mov    DWORD PTR [rax+0x8],0x80208210;   {metadata('hello$A')}
  0x00007f7eccad9c6e:   mov    DWORD PTR [rax+0xc],0x0
```

(Ofc, I may have missed some jvm flags; the corresponding source `PhaseMacroExpand::initialize_object` is not obvious to me.)